scipy/interpolate/tests/test_aaa.py

```diff
@@ -0,0 +1,217 @@
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+import numpy as np
+from numpy.testing import assert_allclose, assert_equal
+import pytest
+import scipy
+from scipy.interpolate import AAA
+
+TOL = 1e4 * np.finfo(np.float64).eps
+UNIT_INTERVAL = np.linspace(-1, 1, num=1000)
+PTS = np.logspace(-15, 0, base=10, num=500)
+PTS = np.concatenate([-PTS[::-1], [0], PTS])
+
+class TestAAA:
+    def test_input_validation(self):
+        with pytest.raises(ValueError, match="same size"):
+            AAA([0], [1, 1])
+        with pytest.raises(ValueError, match="1-D"):
+            AAA([[0], [0]], [[1], [1]])
+    
+    def test_convergence_error(self):
+        with pytest.warns(RuntimeWarning, match="AAA failed"):
+            AAA(UNIT_INTERVAL, np.exp(UNIT_INTERVAL),  max_terms=1)
+
+    @pytest.mark.parametrize("dtype", [np.float32, np.float64, np.complex64, 
+                                       np.complex128])
+    def test_dtype_preservation(self, dtype):
+        rtol = np.finfo(dtype).eps ** 0.75 * 100
+        rng = np.random.default_rng(59846294526092468)
+
+        z = np.linspace(-1, 1, dtype=dtype)
+        r = AAA(z, np.sin(z))
+        
+        z2 = rng.uniform(-1, 1, size=100).astype(dtype)
+        assert_allclose(r(z2), np.sin(z2), rtol=rtol)
+        assert r(z2).dtype == dtype
+        
+        assert r.support_points.dtype == dtype
+        assert r.values.dtype == dtype
+        assert r.weights.dtype == dtype
+        assert r.errors.dtype == z.real.dtype
+        assert r.poles().dtype == np.result_type(dtype, 1j)
+        assert r.residues().dtype == np.result_type(dtype, 1j)
+        assert r.roots().dtype == np.result_type(dtype, 1j)
+    
+    # The following tests are based on:
+    # https://github.com/chebfun/chebfun/blob/master/tests/chebfun/test_aaa.m
+    def test_exp(self):
+        f = np.exp(UNIT_INTERVAL)
+        r = AAA(UNIT_INTERVAL, f)
+
+        assert_allclose(r(UNIT_INTERVAL), f, atol=TOL)
+        assert_equal(r(np.nan), np.nan)
+        assert np.isfinite(r(np.inf))
+
+        m1 = r.support_points.size
+        r = AAA(UNIT_INTERVAL, f, rtol=1e-3)
+        assert r.support_points.size < m1
+
+    def test_tan(self):
+        f = np.tan(np.pi * UNIT_INTERVAL)
+        r = AAA(UNIT_INTERVAL, f)
+
+        assert_allclose(r(UNIT_INTERVAL), f, atol=10 * TOL, rtol=1.4e-7)
+        assert_allclose(np.min(np.abs(r.roots())), 0, atol=3e-10)
+        assert_allclose(np.min(np.abs(r.poles() - 0.5)), 0, atol=TOL)
+        # Test for spurious poles
+        assert np.min(np.abs(r.residues())) > 1e-13
```

Comment:
Double check that this test does what it's supposed to.

Reply:
Have expanded the comment, spurious poles have tiny residues hence the check